jax/_src/nn/functions.py

```diff
@@ -765,3 +768,203 @@ def hard_silu(x: ArrayLike) -> Array:
   return x_arr * hard_sigmoid(x_arr)
 
 hard_swish = hard_silu
+
+def _causal_mask(T, dtype):
+  dtype_max = jnp.finfo(dtype).max
+  large_negative_number = jnp.asarray(-0.5 * dtype_max, dtype=dtype)
+  col_idx = jnp.tile(jnp.arange(T)[jnp.newaxis, :], [T, 1])
+  row_idx = jnp.tile(jnp.arange(T)[:, jnp.newaxis], [1, T])
+  mask = (row_idx < col_idx).astype(dtype) * large_negative_number
+  mask = mask[jnp.newaxis, jnp.newaxis, :, :]
+  return mask
+
+class SdpaCausalMask:
```

Comment:
This name is non-ideal, better to avoid this confusing acronym (`ScaledDotProductAttentionCausalMask` is clearer).